### PR TITLE
Allow specifying file start position

### PR DESCRIPTION
**Pull request by pickfire**
Like helix-term/src/commands.rs:3426:15

### CODE REVIEW

File: helix-view/src/editor.rs [hunk 1]
```diff
@@ -19,6 +19,19 @@ pub use helix_core::register::Registers;
 use helix_core::syntax;
 use helix_core::Position;
 
+/// File path with row and column.
```
Comment (kirawi) [inline helix-view/src/editor.rs:22]: I don't completely understand everything that's going on here, and this is my first time seriously reviewing a PR, but is there a reason that `File` can't be baked into `Document`? Or is this to make the intent of the code clearer?

Comment (pickfire) [reply helix-view/src/editor.rs:22]: Ah, this file is not the normal file, this is the file argument being passed into `:open` or `hx <file>`.

File: helix-term/src/args.rs [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
 #[derive(Default)]
 pub struct Args {
     pub display_help: bool,
     pub display_version: bool,
     pub verbosity: u64,
-    pub files: Vec<PathBuf>,
+    pub files: Files,
+}
+
+pub enum Files {
```
Comment (archseer) [inline helix-term/src/args.rs:13]: I don't like this extra enum.

File: helix-view/src/editor.rs [hunk 3]
```diff
@@ -114,7 +127,7 @@ impl Editor {
         }
     }
 
-    pub fn switch(&mut self, id: DocumentId, action: Action) {
+    pub fn switch(&mut self, id: DocumentId, action: Action, pos: Option<(u32, u32)>) {
```
Comment (archseer) [inline helix-view/src/editor.rs:130]: I don't like the extra parameter, why not just make it `switch` + `jump_to` if there's a position? That's how the goto jumps work across files.

File: helix-view/src/editor.rs [hunk 4]
```diff
@@ -140,10 +153,12 @@ impl Editor {
                 let (view, doc) = current!(self);
 
                 // initialize selection for view
-                let selection = doc
-                    .selections
-                    .entry(view.id)
-                    .or_insert_with(|| Selection::point(0));
+                let pos = if let Some((row, col)) = pos {
```
Comment (archseer) [inline helix-view/src/editor.rs:156]: I think for code like this it's easier to just `pos.unwrap_or_default()` (default ends up being `(0, 0)` and eliminate the two branches.

File: helix-view/src/editor.rs [hunk 5]
```diff
@@ -174,12 +199,12 @@ impl Editor {
...
         let id = self.documents.insert(doc);
         self.documents[id].id = id;
-        self.switch(id, action);
+        self.switch(id, action, None);
         id
     }
 
-    pub fn open(&mut self, path: PathBuf, action: Action) -> Result<DocumentId, Error> {
-        let path = crate::document::canonicalize_path(&path)?;
+    pub fn open(&mut self, file: File, action: Action) -> Result<DocumentId, Error> {
```
Comment (archseer) [inline helix-view/src/editor.rs:206]: Same here, I think adding a new type makes this much more annoying to use than `open` + `jump`.

**Review by archseer — CHANGES_REQUESTED** [review]

File: helix-term/src/args.rs (hunk 2 above)
Comment (pickfire) [reply helix-term/src/args.rs:13]: What alternative do you think there is? I don't know if there is a better way.

File: helix-view/src/editor.rs (hunk 3 above)
Comment (pickfire) [reply helix-view/src/editor.rs]: Fixed

File: helix-view/src/editor.rs (hunk 5 above)
Comment (pickfire) [reply helix-view/src/editor.rs]: Fixed

File: helix-term/src/application.rs [hunk 6]
```diff
@@ -78,20 +85,30 @@ impl Application {
...
                             "expected a path to file, found a directory. (to open a directory pass it as first argument)"
                         ));
                     } else {
-                        editor.open(file.to_path_buf(), Action::Load)?;
+                        editor.open(file, Action::Load)?;
+                        // let docid = editor.open(file, Action::Load)?;
+                        // editor.document_mut(docid).unwrap().set_selection();
+                        // FIXME: current is incorrect since it will only work for current
+                        // document, other files loaded in background will result in panic
+                        let (view, doc) = current!(editor);
```
Comment (pickfire) [inline helix-term/src/application.rs:106]: Not sure how to fix this. It should get the correct view that is not focused.

Comment (archseer) [reply helix-term/src/application.rs:106]: You should do nothing here since the file is loaded in the background and not assigned to a view.

If you want to set a location for the current split, you can push a selection onto the doc `doc.selections.insert(view.id, ...)`

Comment (pickfire) [reply helix-term/src/application.rs:106]: > You should do nothing here since the file is loaded in the background and not assigned to a view.

But the user did specify the file position, so I don't think I should do nothing here.

File: helix-term/src/args.rs [hunk 7]
```diff
@@ -41,15 +41,31 @@ impl Args {
...
+
+/// Parse arg into [`PathBuf`] and position.
+pub(crate) fn parse_file(s: &str) -> (PathBuf, usize, usize) {
+    split_path_pos(s).unwrap_or_else(|| (PathBuf::from(s), 0, 0))
+}
+
+/// Split file.rs:10:2 into [`PathBuf`] and position.
+///
+/// Does not validate if file.rs is a file or directory.
+fn split_path_pos(s: &str) -> Option<(PathBuf, usize, usize)> {
```
Comment (cole-h) [inline helix-term/src/args.rs:65]: Would it be desirable / acceptable to parse `filename:10` as `filename:10:0`? Sometimes, I know what line I want to go to (e.g. by grepping through some directory), but I don't care about the column. Sure, it's just three extra keystrokes, but I, personally, think it would be nice.

Comment (pickfire) [reply helix-term/src/args.rs:65]: `filename:10:0` won't work since it should be `1`, it is one indexed, not zero indexed.

Also, I don't think I know any use case that provide the output in the format of `filename:row`, even with grep it outputs it separately so I don't know if it is useful for it to be here.

Unless there is like something that outputs `filename:row` directly then I think it's worth implementing, otherwise here still needs some effort to type out the filename + pos.

Comment (pickfire) [reply helix-term/src/args.rs]: Fixed

File: helix-term/src/application.rs [hunk 8]
```diff
@@ -116,24 +123,37 @@ impl Application {
...
                         ));
                     } else {
-                        editor.open(file.to_path_buf(), Action::Load)?;
+                        let doc_id = editor.open(file, Action::Load)?;
+                        // with Action::Load all documents have the same view
+                        let view_id = editor.tree.focus;
+                        let doc = editor.document_mut(doc_id).unwrap();
+                        let row = row.min(doc.text().len_lines() - 1);
+                        let pos = Selection::point(
+                            (doc.text().line_to_char(row) + col)
```
Comment (cole-h) [inline helix-term/src/application.rs:146]: `ropey` exposes a fallible version of `line_to_char` as `try_line_to_char` -- maybe you could use that and if it fails, unwrap to the default selection at `(0,0)`?

If you try to open a file to an obviously incorrect line (e.g. `hx README.md:100000000:1`), helix will panic because it's obviously out of bounds.

Comment (cole-h) [reply helix-term/src/application.rs:146]: I came up with a minor diff that would implement this functionality (but instead of treating an over-long index as 0, it just goes to the end of the file, which I think is more correct):

```diff
diff --git a/helix-term/src/application.rs b/helix-term/src/application.rs
index 2a86cff..c5ab2aa 100644
--- a/helix-term/src/application.rs
+++ b/helix-term/src/application.rs
@@ -1,4 +1,7 @@
-use helix_core::{line_ending::line_end_char_index, merge_toml_values, syntax, Selection};
+use helix_core::{
+    line_ending::{line_end_char_index, rope_end_without_line_ending},
+    merge_toml_values, syntax, Selection,
+};
 use helix_lsp::{lsp, util::lsp_pos_to_pos, LspProgressMap};
 use helix_view::{theme, Editor};
 
@@ -141,11 +144,14 @@ impl Application {
                         // with Action::Load all documents have the same view
                         let view_id = editor.tree.focus;
                         let doc = editor.document_mut(doc_id).unwrap();
-                        let pos = Selection::point(
-                            (doc.text().line_to_char(row) + col)
-                                .min(line_end_char_index(&doc.text().slice(..), row)),
-                        );
-                        doc.set_selection(view_id, pos);
+                        let selection = if let Ok(pos) = doc.text().try_line_to_char(row) {
+                            Selection::point(
+                                (pos + col).min(line_end_char_index(&doc.text().slice(..), row)),
+                            )
+                        } else {
+                            Selection::point(rope_end_without_line_ending(&doc.text().slice(..)))
+                        };
+                        doc.set_selection(view_id, selection);
                     }
                 }
                 editor.set_status(format!("Loaded {} files.", nr_of_files));

```

Comment (pickfire) [reply helix-term/src/application.rs:146]: > If you try to open a file to an obviously incorrect line (e.g. hx README.md:100000000:1), helix will panic because it's obviously out of bounds.

I tested this but it doesn't panic. I did fixed the panics previously already.

Comment (cole-h) [reply helix-term/src/application.rs:146]: Oh, indeed, looks like I was on an old commit. Sorry for the noise!

File: helix-term/src/args.rs [hunk 9]
```diff
@@ -41,15 +41,43 @@ impl Args {
...
+pub(crate) fn parse_file(s: &str) -> (PathBuf, usize, usize) {
+    split_path_row_col(s)
+        .or_else(|| split_path_row(s))
+        .unwrap_or_else(|| (PathBuf::from(s), 0, 0))
+}
+
+/// Split file.rs:10:2 into [`PathBuf`], row and col.
+///
+/// Does not validate if file.rs is a file or directory.
+fn split_path_row_col(s: &str) -> Option<(PathBuf, usize, usize)> {
```
Comment (archseer) [inline helix-term/src/args.rs:67]: This should use `Option<(PathBuf, Position)>`: https://github.com/helix-editor/helix/blob/cffd0f8a6a02cf153159be608a7d1b3669a92bc4/helix-core/src/position.rs

Comment (pickfire) [reply helix-term/src/args.rs]: Thanks, I forgot about that.

File: helix-term/src/commands.rs [hunk 10]
```diff
@@ -1767,8 +1768,15 @@ mod cmd {
         args: &[&str],
         _event: PromptEvent,
     ) -> anyhow::Result<()> {
-        let path = args.get(0).context("wrong argument count")?;
-        let _ = cx.editor.open(path.into(), Action::Replace)?;
+        use helix_core::path::expand_tilde;
+        let arg = args.get(0).context("wrong argument count")?;
+        let (path, pos) = args::parse_file(arg);
+        let _ = cx.editor.open(expand_tilde(&path), Action::Replace)?;
```
Comment (archseer) [inline helix-term/src/commands.rs:1774]: There's no need to call `expand_tilde` anymore, the underlying bug got fixed

```suggestion
        let arg = args.get(0).context("wrong argument count")?;
        let (path, pos) = args::parse_file(arg);
        let _ = cx.editor.open(path.into(), Action::Replace)?;
```

File: helix-core/src/position.rs [hunk 11]
```diff
@@ -109,7 +109,10 @@ pub fn visual_coords_at_pos(text: RopeSlice, pos: usize, tab_width: usize) -> Po
 /// TODO: this should be changed to work in terms of visual row/column, not
 /// graphemes.
 pub fn pos_at_coords(text: RopeSlice, coords: Position, limit_before_line_ending: bool) -> usize {
-    let Position { row, col } = coords;
+    let Position { mut row, col } = coords;
+    if limit_before_line_ending {
+        row = row.min(text.len_lines() - 1);
+    };
```
Comment (archseer) [inline helix-core/src/position.rs:115]: Is this really necessary? It's handled a few lines lower

**Review by archseer — CHANGES_REQUESTED** [review]

**Comment by Curiosidad-Racional** [comment]
Linux allows `:` in the file name. With this feature, how to open a file named `helix-term/src/commands.rs:3426:15` from the command line?

**Comment by cole-h** [comment]
It works normally:

```
cargo r -- test:10:11:2
```

where the filename is `test:10` properly brings me to line 11, column 2.

**Comment by Curiosidad-Racional** [comment]
> EDIT: Oh, I see your point now.
> 
> Then I suppose we should probably check if there's a file with the same name before starting to parse it for line and column.

Then... How to open a file named `helix-term/src/commands.rs:3426:15` with other file named `helix-term/src/commands.rs`?

**Comment by cole-h** [comment]
There's no perfect solution. Even for your PR of `+line:column`, what if I have a file named exactly `+10:10`? How do I open file `asdf` first, alongside `+10:10` without going to line 10 and column 10 of file `asdf`?

Do you find yourself editing files named `test:10:10` alongside a file named `test` that often? To me, this sounds like an edge case that should be acknowledged, but not supported.

IMHO, I'd much rather support this syntax because that's what almost every compiler outputs in their diagnostics.

**Comment by Curiosidad-Racional** [comment]
> There's no perfect solution. Even for your PR of `+line:column`, what if I have a file named exactly `+10:10`? How do I open file `asdf` first, alongside `+10:10` without going to line 10 and column 10 of file `asdf`?

Linux way: 'hx -- +10:10` opens a file named `+10:10`.

**Comment by pickfire** [comment]
I am aware of this but I haven't encounter it myself.

@Curiosidad-Racional please choose better wording, you made it sound like I purposely write this patch to annoy users.

My original thoughts was check if it exists before splitting, but I think since file extension is usually at the back rather than the syntax here so I think it won't be an issue.

File: helix-core/src/position.rs (hunk 11 above)
Comment (pickfire) [reply helix-core/src/position.rs:115]: Yes, or otherwise it will panic.

**Comment by Curiosidad-Racional** [comment]
> I am aware of this but I haven't encounter it myself.
> 
> @Curiosidad-Racional please choose better wording, you made it sound like I purposely write this patch to annoy users.
As you may have noticed I'm not a native English speaker. My intention was not to offend but to clarify. I was referring to annoying behaviors:

1. opening a file named `file.log2021-11-21_08:03:01`
2. creating a file named `file.log2021-11-21_08:03:01`

I didn't refer to the PR. Altruistic contributions are admirable.

> 
> My original thoughts was check if it exists before splitting, but I think since file extension is usually at the back rather than the syntax here so I think it won't be an issue.
IMO users espect `hx [filename]` to open literal `[filename]`. Other behaviors are usually with modifiers `hx --compiler <filename:line:col>`. This is an example, I don't pretend that it is implemented like this... I'm giving ideas not solutions.

> 
> Or maybe I should just add the check before splitting it?
Whatever you prefer. We just don't make a possible use of `helix` impossible (1. or 2.).
 
:-)

**Comment by pickfire** [comment]
I implemented a check before splitting.

File: helix-term/src/args.rs [hunk 12]
```diff
@@ -41,15 +42,51 @@ impl Args {
...
+        .unwrap_or_else(|| (PathBuf::from(s), Position::default()))
+}
+
+/// Split file.rs:10:2 into [`PathBuf`], row and col.
+///
+/// Does not validate if file.rs is a file or directory.
+fn split_path_row_col(s: &str) -> Option<(PathBuf, Position)> {
+    if Path::new(s).exists() {
+        return None;
+    }
```
Comment (cole-h) [inline helix-term/src/args.rs:71]: Seems like this should go in `parse_file` instead to limit duplication of code.

File: helix-term/src/main.rs [hunk 13]
```diff
@@ -56,7 +56,7 @@ USAGE:
     hx [FLAGS] [files]...
 
 ARGS:
-    <files>...    Sets the input file to use
+    <files>...    Sets the input file to use or with pos file[:row[:col]]
```
Comment (archseer) [inline helix-term/src/main.rs:59]:
```suggestion
    <files>...    Sets the input file to use, position can also be specified via file[:row[:col]]
```

File: helix-term/src/args.rs (hunk 12 above)
Comment (pickfire) [reply helix-term/src/args.rs:71]: Thanks, I missed this.

File: helix-term/src/main.rs (hunk 13 above)
Comment (pickfire) [reply helix-term/src/main.rs]: Screenshot_20211129_225244

**Review by archseer — COMMENTED** [review]
Looks like there's a few merge conflicts

File: helix-term/src/args.rs [hunk 14]
```diff
@@ -41,15 +42,49 @@ impl Args {
...
+    let path = s.next()?.into();
+    let pos = Position::new(row.saturating_sub(1), col.saturating_sub(1));
+    Some((path, pos))
+}
+
+/// Split file.rs:10 into [`PathBuf`] and row.
+///
+/// Does not validate if file.rs is a file or directory.
+fn split_path_row(s: &str) -> Option<(PathBuf, Position)> {
+    let mut s = s.rsplitn(2, ':');
```
[the remark this answers is not in the document]
Comment (pickfire) [reply helix-term/src/args.rs:85]: `rsplit_last`? Where can I find that?

[the remark this answers is not in the document]
Comment (archseer) [reply helix-term/src/args.rs:85]: Sorry, `rsplit_once` https://doc.rust-lang.org/std/primitive.str.html#method.rsplit_once

[the remark this answers is not in the document]
Comment (pickfire) [reply helix-term/src/args.rs:85]: Nice, I didn't know that.
